revolutionpi/analog_pin.go

```diff
@@ -17,101 +23,106 @@ type analogPin struct {
+	analogPin.outputOffset = aio.i16uOutputOffset
+	analogPin.inputOffset = aio.i16uInputOffset
+
+	if analogPin.isAnalogInput() {
+		analogInputNumber := (analogPin.Address - analogPin.inputOffset) / 2                     // results in 0, 1, 2, or 3
+		inputRangeAddress := analogInputNumber*7 + analogInputMemAddress + analogPin.inputOffset // results in pin 24, 31, 38, or 45
+		bufInputRange := make([]byte, 1)
+		n, err := analogPin.ControlChip.fileHandle.ReadAt(bufInputRange, int64(inputRangeAddress))
+		if err != nil {
+			return nil, fmt.Errorf("failed to read input range for analog pin %s", analogPin.Name)
+		}
+		if n != 1 {
+			return nil, fmt.Errorf("expected 1 byte, got %#v", bufInputRange)
+		}
+		analogPin.info, err = getAnalogInputRange(bufInputRange[0])
+		if err != nil {
+			return nil, err
+		}
+	} else if analogPin.isAnalogOutput() {
+		// check to see if analog output is enabled
+		outputRangeAddress := analogPin.inputOffset + 69
+		// use the corresponding analog OutputRange pin to check if the analog output is enabled
+		if analogPin.Address == analogPin.outputOffset+2 {
+			outputRangeAddress = analogPin.inputOffset + 79
+		}
+		bufOutputRange := make([]byte, 1)
+		n, err := analogPin.ControlChip.fileHandle.ReadAt(bufOutputRange, int64(outputRangeAddress))
+		if err != nil {
+			return nil, err
+		}
+		if n != 1 {
+			return nil, fmt.Errorf("unable to determine if pin %s is configured for analog write", analogPin.Name)
+		}
+		analogPin.ControlChip.logger.Debugf("outputRange Value: %d", bufOutputRange)
+		analogPin.info, err = getAnalogOutputRange(bufOutputRange[0], analogPin.Name)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &analogPin, nil
 }
 
-func (pin *analogPin) Read(ctx context.Context, extra map[string]interface{}) (int, error) {
+func (pin *analogPin) Read(ctx context.Context, extra map[string]interface{}) (board.AnalogValue, error) {
 	if !pin.isAnalogInput() {
-		return 0, fmt.Errorf("cannot ReadAnalog, pin %s is not an analog input pin", pin.Name)
+		return board.AnalogValue{}, fmt.Errorf("cannot ReadAnalog, pin %s is not an analog input pin", pin.Name)
 	}
 	pin.ControlChip.logger.Debugf("Reading from %v, length: %v byte(s)", pin.Address, pin.Length/8)
 	b := make([]byte, pin.Length/8)
 	n, err := pin.ControlChip.fileHandle.ReadAt(b, int64(pin.Address))
 	pin.ControlChip.logger.Debugf("Read %#v bytes", b)
 	if n != 2 {
-		return 0, fmt.Errorf("expected 2 bytes, got %#v", b)
+		return board.AnalogValue{}, fmt.Errorf("expected 2 bytes, got %#v", b)
 	}
 	if err != nil {
-		return 0, err
+		return board.AnalogValue{}, err
 	}
 	val := binary.LittleEndian.Uint16(b)
-	return int(val), nil
+	// NOTE: we currently assume that the input multiplier, divisor, and offset have not been modified
+	// the min and max values will change if a user modifies these.
+	analogVal := board.AnalogValue{Value: int(val), Min: float32(pin.info.min), Max: float32(pin.info.max), StepSize: 0.001}
```

Comment:
using `StepSize: 0.001` since the rev pi returns in mV and not a N-bit digital value. Unsure if this is the right way to go about this tho